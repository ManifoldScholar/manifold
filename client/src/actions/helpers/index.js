export { default as createApiAction } from "./createApiAction";
export { default as commonActions } from "./commonActions";
