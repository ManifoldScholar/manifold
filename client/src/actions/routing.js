import { createAction } from "redux-actions";

export const update = createAction("ROUTE_UPDATE", state => state);
