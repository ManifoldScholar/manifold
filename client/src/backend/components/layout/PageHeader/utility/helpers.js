import lh from "helpers/linkHandler";

export const getTextLinks = ({ texts, pathname }) =>
  texts?.length
    ? texts?.map(t => ({
        label: t.label,
        route: "backendTextAnalytics",
        id: t.id,
        active: pathname?.includes(t.id)
      }))
    : null;

export const getIssueLinks = ({ issues, pathname }) =>
  issues?.length
    ? issues?.map(i => ({
        label: i.label,
        route: "backendProjectAnalytics",
        id: i.id,
        active: pathname?.includes(i.id)
      }))
    : null;

export const getLinkOrButtonProps = action => {
  if (action.href) {
    return action.download
      ? { as: "a", href: action.href, download: true }
      : {
          as: "a",
          href: action.href,
          target: "_blank",
          re: "noopener noreferrer"
        };
  }

  if (action.route)
    return {
      to: lh.link(action.route, action.slug, action.resourceSlug)
    };

  return { as: "button", onClick: action.onClick };
};
