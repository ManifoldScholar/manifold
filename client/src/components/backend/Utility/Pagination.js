import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import classNames from 'classnames';

export default class UtilityPagination extends Component {

  static displayName = "Utility.Pagination";

  static propTypes = {
  };

  constructor() {
    super();
  }

  render() {
    return (
      <nav className="list-pagination-secondary">
        <div className="pagination-previous">
          <Link to="#">
            <i className="manicon manicon-arrow-long-left"></i>
            Prev
          </Link>
        </div>
        <div className="pagination-ordinals">
          {'Page 1 / 4'}
        </div>
        <div className="pagination-next">
          <Link to="#">
            Next
            <i className="manicon manicon-arrow-long-right"></i>
          </Link>
        </div>
      </nav>
    );
  }

}
