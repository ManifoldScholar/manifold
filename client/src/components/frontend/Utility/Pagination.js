import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import classNames from 'classnames';

export default class UtilityPagination extends Component {

  static displayName = "Utility.Pagination";

  static propTypes = {
  };

  constructor() {
    super();
  }

  render() {
    return (
      <nav className="list-pagination">
        <ul>
          <li className="pagination-previous">
            <Link to="#">
              <i className="manicon manicon-arrow-long-left"></i>
              Prev
            </Link>
          </li>
          <li>
            <Link to="#">1</Link>
          </li>
          <li className="active">
            <Link to="#">2</Link>
          </li>
          <li>
            <Link to="#">3</Link>
          </li>
          <li>
            <Link to="#">4</Link>
          </li>
          <li>
            <Link to="#">5</Link>
          </li>
          <li className="pagination-next">
            <Link to="#">
              Next
              <i className="manicon manicon-arrow-long-right"></i>
            </Link>
          </li>
        </ul>
      </nav>
    );
  }

}
