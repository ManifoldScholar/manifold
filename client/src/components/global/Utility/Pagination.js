import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import range from "lodash/range";
import isString from "lodash/isString";

export default class UtilityPagination extends PureComponent {
  static displayName = "Utility.Pagination";

  static propTypes = {
    pagination: PropTypes.object,
    paginationTarget: PropTypes.string,
    padding: PropTypes.number,
    paginationClickHandler: PropTypes.func
  };

  static defaultProps = {
    padding: 3,
    paginationTarget: "#pagination-target"
  };

  visiblePageArray(pagination) {
    const current = pagination.currentPage;
    let start = current - this.props.padding;
    if (start < 1) start = 1;
    let last = current + this.props.padding;
    if (last > pagination.totalPages) last = pagination.totalPages;
    const pages = range(start, last + 1);
    const out = pages.map(page => {
      return {
        number: page,
        key: page,
        current: page === current,
        first: page === 1,
        last: page === pagination.totalPages
      };
    });
    return out;
  }

  previous(pagination) {
    const handler = this.props.paginationClickHandler(pagination.prevPage);
    const disabledClass = pagination.prevPage ? "" : "disabled";
    return (
      <li className={`pagination-previous ${disabledClass}`} key="previous">
        {isString(handler) ? (
          <Link to={handler}>
            <i className="manicon manicon-arrow-long-left" />
            <span>Prev</span>
          </Link>
        ) : (
          <a
            href={this.props.paginationTarget}
            onClick={handler}
            data-id={"page-prev"}
          >
            <i className="manicon manicon-arrow-long-left" />
            <span>Prev</span>
          </a>
        )}
      </li>
    );
  }

  next(pagination) {
    const handler = this.props.paginationClickHandler(pagination.nextPage);
    const disabledClass = pagination.nextPage ? "" : "disabled";
    return (
      <li className={`pagination-next ${disabledClass}`} key="next">
        {isString(handler) ? (
          <Link to={handler}>
            <span>Next</span>
            <i className="manicon manicon-arrow-long-right" />
          </Link>
        ) : (
          <a
            href={this.props.paginationTarget}
            onClick={handler}
            data-id={"page-next"}
          >
            <span>Next</span>
            <i className="manicon manicon-arrow-long-right" />
          </a>
        )}
      </li>
    );
  }

  number(page, handler) {
    return (
      <li key={page.key} className="ordinal">
        {isString(handler) ? (
          <Link to={handler}>{page.number}</Link>
        ) : (
          <a href={this.props.paginationTarget} onClick={handler}>
            {page.number}
          </a>
        )}
      </li>
    );
  }

  current(page, handler) {
    return (
      <li className="active ordinal" key={page.key}>
        {isString(handler) ? (
          <Link to={handler}>{page.number}</Link>
        ) : (
          <a href={this.props.paginationTarget} onClick={handler}>
            {page.number}
          </a>
        )}
      </li>
    );
  }

  render() {
    if (!this.props.pagination) return null;

    const pages = this.visiblePageArray(this.props.pagination);
    const pagination = this.props.pagination;
    if (pagination.totalPages === 1 || pagination.totalPages === 0) return null;

    return (
      <nav className="list-pagination-primary">
        <ul>
          {this.previous(pagination)}
          {pages.map(page => {
            const url = this.props.paginationClickHandler(page.number);
            if (page.current) {
              return this.current(page, url);
            }
            return this.number(page, url);
          })}
          {this.next(pagination)}
        </ul>
      </nav>
    );
  }
}
