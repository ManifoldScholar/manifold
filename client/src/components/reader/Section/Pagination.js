import React, { Component } from 'react';
import PropTypes from 'prop-types';
import lh from 'helpers/linkHandler';
import { Link } from 'react-router-dom';

export default class Pagination extends Component {

  static propTypes = {
    textId: PropTypes.string,
    sectionId: PropTypes.string,
    spine: PropTypes.array
  };

  getSiblingSection(id, shift) {
    if (!this.props.spine) return;
    let siblingSection = false;
    const index = this.props.spine.indexOf(id);
    if (this.props.spine[index + shift] || index !== -1) {
      siblingSection = this.props.spine[index + shift];
    }

    return siblingSection;
  }

  getSectionPath(id) {
    return lh.link("readerSection", this.props.textId, id);
  }

  getPreviousLink() {
    let previousLink = null;
    const previousNode = this.getSiblingSection(this.props.sectionId, - 1);
    if (previousNode) {
      const previousPath = this.getSectionPath(previousNode);
      previousLink = (
        <Link to={previousPath} className="pagination-previous" >
          <i className="manicon manicon-arrow-round-left"></i>
          <span className="text">
            {'Previous'}
          </span>
        </Link>
      );
    }
    return previousLink;
  }

  getNextLink() {
    let nextLink = null;
    const nextNode = this.getSiblingSection(this.props.sectionId, 1);
    if (nextNode) {
      const nextPath = this.getSectionPath(nextNode);
      nextLink = (
        <Link to={nextPath} className="pagination-next" >
          <span className="text">
            {'Next'}
          </span>
          <i className="manicon manicon-arrow-round-right"></i>
        </Link>
      );
    }
    return nextLink;
  }

  render() {
    return (
      <nav className="section-pagination">
        <div className="container">
          {this.getPreviousLink()}
          {this.getNextLink()}
        </div>
      </nav>
    );
  }
}
