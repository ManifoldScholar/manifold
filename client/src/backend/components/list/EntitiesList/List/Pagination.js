import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Utility from "global/components/utility";

export default class ListEntitiesListPagination extends PureComponent {
  static displayName = "List.Entities.List.Pagination";

  static propTypes = {
    pagination: PropTypes.object.isRequired,
    onPageClick: PropTypes.func.isRequired,
    paginationTarget: PropTypes.string,
    usesQueryParams: PropTypes.bool,
    style: PropTypes.oneOf(["normal", "compact"]),
    padding: PropTypes.number
  };

  static defaultProps = {
    style: "normal"
  };

  get pagination() {
    return this.props.pagination;
  }

  get style() {
    return this.props.style;
  }

  get styleProps() {
    if (this.style === "compact") {
      return { compact: true };
    }
    return { compact: false, padding: this.props.padding };
  }

  get onPageClick() {
    return this.props.onPageClick;
  }

  get paginationTarget() {
    if (this.props.usesQueryParams) return undefined;

    return this.props.paginationTarget ?? "#";
  }

  get padding() {
    return this.props.padding;
  }

  render() {
    return (
      <div className="entity-list__pagination">
        <Utility.Pagination
          pagination={this.pagination}
          paginationTarget={this.paginationTarget}
          paginationClickHandler={this.onPageClick}
          {...this.styleProps}
        />
      </div>
    );
  }
}
