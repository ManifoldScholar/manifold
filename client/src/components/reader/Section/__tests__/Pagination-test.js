import React from "react";
import Pagination from "../Pagination";
import renderer from "react-test-renderer";
import { wrapWithRouter, renderWithRouter } from "test/helpers/routing";
import build from "test/fixtures/build";

describe("Reader.Section.Pagination component", () => {
  const text = build.entity.text("1");
  const root = wrapWithRouter(
    <Pagination text={text} sectionId={"2"} spine={["1234", "5678", "0000"]} />
  );

  it("renders correctly", () => {
    const component = renderer.create(root);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
