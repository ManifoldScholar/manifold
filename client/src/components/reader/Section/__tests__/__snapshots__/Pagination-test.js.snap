// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader.Section.Pagination component renders correctly 1`] = `
<nav
  className="section-pagination"
>
  <div
    className="container"
  >
    <a
      className="pagination-next"
      href="/read/1/section/1234"
      onClick={[Function]}
    >
      <span
        className="text"
      >
        Next
      </span>
      <i
        className="manicon manicon-arrow-round-right"
      />
    </a>
  </div>
</nav>
`;
